### PR TITLE
Added Form Event Listener to set name of root taxon

### LINKED ISSUE #2734: Link to view products by taxon is throwing No route found Exception
- Fresh install of Sylius (commit: c38485c581007f8f29fd8ba9f02b6d55537d920c). 
- Ran sylius:install and sylius:install:setup but did not load the sample data
- Logged into admin area and created one taxonomy with one root taxon and one child taxon
- Created one product belonging to the child taxon

Expected:
- When I click on the product link from homepage I see the product
- When I click on the child taxon in homepage I see the product

Actual:
- When I click on the product link from homepage I see the product
- When I click on the child taxon in homepage I get the following error:
 No route found for "GET /t//flooring/wood" (from "http://127.0.0.1:8000/app_dev.php/")
404 Not Found - NotFoundHttpException
1 linked Exception:

    ResourceNotFoundException » 
[1/2] ResourceNotFoundException: None of the routers in the chain matched this request
GET /app_dev.php/t//flooring/wood HTTP/1.1

I'm not a Symfony expert but I can see the relevant routes defined in vendor/sylius/sylius/src/Sylius/Bundle/CoreBundle/Resources/config/app/sylius.yml
It looks like it's finding the %sylius.model.product.class% route but not the %sylius.model.taxon.class% route.

Any ideas why? Happy to submit a PR if it is an issue and you can point me in the right direction.

### PR DESCRIPTION
**Pull request by liquorvicar**
Fixes #2734 
Not a Symfony expert so not sure if this the best way to solve this problem or whether this will still leave an issue elsewhere?

Updated one relevant spec but let me know if this needs more specs or behat features.

### CODE REVIEW

File: src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/SetRootTaxonNameFormListener.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            FormEvents::POST_SUBMIT  => 'postBind'
+        );
+    }
+
+    /**
+     * Make sure we set the name of the root taxon
+     *
+     * @param FormEvent $event
+     */
+    public function postBind(FormEvent $event)
```
Comment (aRn0D) [inline src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/SetRootTaxonNameFormListener.php]: postSubmit should be better ;)

File: src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/SetRootTaxonNameFormListener.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Form\FormFactoryInterface;
+
+/**
+ * Adds the taxonomy name to the root taxon
+ *
+ * @author Andy Clyde <me@andyclyde.com>
+ */
+class SetRootTaxonNameFormListener implements EventSubscriberInterface
```
Comment (aRn0D) [inline src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/SetRootTaxonNameFormListener.php]: It is a subscriber, so, you should find a generic name like `TaxonSubscriber`

File: src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php [hunk 3]
```diff
@@ -32,6 +33,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'form_type' => 'sylius_taxonomy_translation',
                 'label'    => 'sylius.form.taxonomy.name'
             ))
+            ->addEventSubscriber(new SetRootTaxonNameFormListener())
```
Comment (aRn0D) [inline src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php]: You should define it as a service

File: src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/SetRootTaxonNameFormListener.php (hunk 2 above)
Comment (aRn0D) [inline src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/SetRootTaxonNameFormListener.php]: You should create a spec for that class

Comment (liquorvicar) [inline src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/SetRootTaxonNameFormListener.php]: I followed the naming schema I found in \Sylius\Bundle\TaxonomyBundle\Form\EventListener\BuildTaxonFormListener. Is there a convention in the project for how these classes should be named? Would it be best to call it \Sylius\Bundle\TaxonomyBundle\Form\EventListener\BuildTaxonomyFormListener so it matches the other class in the same folder?

File: src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php (hunk 3 above)
Comment (liquorvicar) [inline src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php]: Would that be something like adding:
```
        <service id="sylius.form.subscriber.build_taxonomy_form" class="Sylius\Bundle\TaxonomyBundle\Form\EventListener\BuildTaxonomyFormListener">
            <tag name="form.subscriber" alias="sylius_taxonomy_form" />
        </service>```
to src/Sylius/Bundle/TaxonomyBundle/Resources/config/services.xml ?
If so, how would I then reference the service in src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php ?

**Comment by liquorvicar** [comment]
@aRn0D Have pushed some changes as per your comments. I still have a question about how we make the Listener into a service. Plus also, whether we need to turn it into a service? Currently it's only used in one place - the \Sylius\Bundle\TaxonomyBundle\Form\Type\TaxonomyType::buildForm method.

**Comment by liquorvicar** [comment]
I've been looking at the code in a bit more detail and it looks like the issue arises because the Symfony form handling is set the name on a taxonomy ny adding a translation to the Taxonomy object and hence bypassing the setName method altogether. Would it be better to add an "addTranslation" method to the Taxonomy class which handles delegating setting the name on the root taxon as well? Is that clearer? i.e. keeps all the behaviour relating to a Taxonomy and how it handles its root Taxon together, rather than splitting some of it out into an Event Listener.

File: src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/BuildTaxonomyFormListener.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Form\FormFactoryInterface;
+
+/**
+ * Adds the taxonomy name to the root taxon
+ *
+ * @author Andy Clyde <me@andyclyde.com>
+ */
+class BuildTaxonomyFormListener implements EventSubscriberInterface
```
Comment (aRn0D) [inline src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/BuildTaxonomyFormListener.php]: it is a Subscriber ;)

File: src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php [hunk 5]
```diff
@@ -32,6 +33,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'form_type' => 'sylius_taxonomy_translation',
                 'label'    => 'sylius.form.taxonomy.name'
             ))
+            ->addEventSubscriber(new BuildTaxonomyFormListener())
```
Comment (aRn0D) [inline src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php]: You shoud define `BuildTaxonomyFormListener` as a service and inject it ino that class. You need to create a constructor and update the form definiftion.

File: src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/BuildTaxonomyFormListener.php (hunk 4 above)
Comment (liquorvicar) [inline src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/BuildTaxonomyFormListener.php]: Well, yes, it does implement the EventSubscriberInterface. But then it's in a directory called \Sylius\Bundle\TaxonomyBundle\Form\EventListener and the other class in that dir is \Sylius\Bundle\TaxonomyBundle\Form\EventListener\BuildTaxonFormListener so I was trying to keep it consistent! Should we rename the whole dir and both classes? Or just leave it as is?

File: src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php (hunk 5 above)
Comment (liquorvicar) [inline src/Sylius/Bundle/TaxonomyBundle/Form/Type/TaxonomyType.php]: Ok,  I have pushed a new commit doing what I think is necessary for this. Can you clarify the logic behind adding this as a service? I based the initial PR on \Sylius\Bundle\TaxonomyBundle\Form\EventListener\BuildTaxonFormListener which doesn't seem to be defined as a service anywhere that I can see.

**Comment by liquorvicar** [comment]
@aRn0D Did you see this comment https://github.com/Sylius/Sylius/pull/2740#issuecomment-103113359 ?

File: src/Sylius/Bundle/TaxonomyBundle/spec/Form/Type/TaxonomyTypeSpec.php [hunk 6]
```diff
@@ -40,6 +41,13 @@ function it_is_a_form_type()
     function it_builds_form_with_proper_fields(FormBuilder $builder)
     {
         $builder
+            ->addEventSubscriber(
+                Argument::type('Sylius\Bundle\TaxonomyBundle\Form\EventListener\BuildTaxonomyFormListener')
+            )
+            ->willReturn($builder)
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/TaxonomyBundle/spec/Form/Type/TaxonomyTypeSpec.php]: Add ``` shouldBeCalled ``` to ensure the method was invoked.

File: src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/BuildTaxonomyFormListener.php (hunk 4 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/TaxonomyBundle/Form/EventListener/BuildTaxonomyFormListener.php]: Keep it consistent and leave it for now, we will update all names at once later.

**Comment by michalmarcinkowski** [comment]
@liquorvicar Would be great if you could cover it with Behat scenario to prevent regression. 

Btw. Great job! :+1:

**Comment by liquorvicar** [comment]
@michalmarcinkowski Thanks for the thumbs :smiley_cat: 

I've been thinking about this issue and I'm not sure I've solved it the best way. I'm slightly concerned that adding it as a form listener means that the data won't get updated correctly if the Model is used elsewhere (the setName() method does the right thing but what if someone adds/updates a translation directly). I have added another PR for what I think might be a better solution: https://github.com/Sylius/Sylius/pull/2786 I don't know too much about the whole translation layer so not sure if what I've done is against how that works.

Would be interested to hear any views on this PR vs https://github.com/Sylius/Sylius/pull/2786.

If we decide to go with this one, will happily add a Behat scenario. Not sure what scenario I would write as the only way I detected this bug manually was by adding a taxonomy/taxon from the backend then clicking on the taxon from the frontend. Could it be something like this:
```
    Scenario: New taxonomy sets root taxon name
        Given that taxonomy "Category" doesn't exist
         When I create a taxonomy called "Category"
         Then the taxonomy called "Category" has a root taxon called "Category"
```

**Comment by pjedrzejewski** [comment]
That was my thinking as well... Forms solve the problem only in forms. Maybe we could use doctrine listener? What do you guys think?

**Comment by liquorvicar** [comment]
A Doctrine listener might work but that still ties that bit of model logic to Doctrine. If you see my other PR I have done it by overriding the addTranslation() method. That feels like a better way to approach it (i.e. keeping the logic purely in the model) but still doesn't quite feel right (plus currently on that PR a load of Behat scenarios are failing so that would need fixing before merging).

**Comment by liquorvicar** [comment]
@pjedrzejewski @michalmarcinkowski 
Any more thoughts on this? Did you see my alternative PR https://github.com/Sylius/Sylius/pull/2786 ?

**Pull request closed without merging**